jre_emul/android/platform/libcore/ojluni/src/main/java/java/net/Inet6Address.java

```diff
@@ -138,138 +142,320 @@
+            ipaddress = new byte[INADDRSZ];
+        }
 
-    /**
-     * This will be set to true when the scope_id field contains a valid
-     * integer scope_id.
-     */
-    boolean scope_id_set = false;
+        private Inet6AddressHolder(
+            byte[] ipaddress, int scope_id, boolean scope_id_set,
+            NetworkInterface ifname, boolean scope_ifname_set)
+        {
+            this.ipaddress = ipaddress;
+            this.scope_id = scope_id;
+            this.scope_id_set = scope_id_set;
+            this.scope_ifname_set = scope_ifname_set;
+            this.scope_ifname = ifname;
+        }
 
-    /**
-     * scoped interface. scope_id is derived from this as the scope_id of the first
-     * address whose scope is the same as this address for the named interface.
-     */
-    @Weak
-    transient NetworkInterface scope_ifname = null;
-
-    // NetworkInterface has two strong references to InetAddress:
-    //  1. NetworkInterface.addrs[] -> InetAddress
-    //  2. NetworkInterface.bindings[] -> InterfaceAddress.address -> InetAddress.
-    // And this class which is a subclass of InetAddress has a field of type NetworkInterface. To
-    // resolve this reference cycle, we can either use the RETAINED_WITH_CHILD_NUM_REFS macro and
-    // set NUM_REFS to 2 or cache all the network interfaces in a map with String keys since the
-    // number of network interfaces on an iOS device is small.
-    /*-[
-    RETAINED_WITH_CHILD_NUM_REFS(scope_ifname_, 2)
-    ]-*/
+        /**
+         * Holds a 128-bit (16 bytes) IPv6 address.
+         */
+        byte[] ipaddress;
+
+        /**
+         * scope_id. The scope specified when the object is created. If the object
+         * is created with an interface name, then the scope_id is not determined
+         * until the time it is needed.
+         */
+        int scope_id;  // 0
+
+        /**
+         * This will be set to true when the scope_id field contains a valid
+         * integer scope_id.
+         */
+        boolean scope_id_set;  // false
+
+        /**
+         * scoped interface. scope_id is derived from this as the scope_id of the first
+         * address whose scope is the same as this address for the named interface.
+         */
+        @Weak
+        NetworkInterface scope_ifname;  // null
+
+        // NetworkInterface has two strong references to InetAddress:
+        //  1. NetworkInterface.addrs[] -> InetAddress
+        //  2. NetworkInterface.bindings[] -> InterfaceAddress.address -> InetAddress.
+        // And this class which is a subclass of InetAddress has a field of type NetworkInterface. To
+        // resolve this reference cycle, we can either use the RETAINED_WITH_CHILD_NUM_REFS macro and
+        // set NUM_REFS to 2 or cache all the network interfaces in a map with String keys since the
+        // number of network interfaces on an iOS device is small.
+        /*-[
+        RETAINED_WITH_CHILD_NUM_REFS(scope_ifname_, 2)
```

Comment:
The macro is moved from an outer class field to an inner class field, thus may need to test for potential retain cycle.

Reply:
You should test for cycles by running "make -f java.mk find_cycles".